# A pilot trial of perioperative IDH Inhibition in treatment-naïve IDH-mutant glioma 

Katharine J. Drummond<sup>1,2,#</sup>, Montana Spiteri<sup>3</sup>, Sarah A. Cain<sup>1</sup>, Jordan Jones<sup>1,2</sup>, Sammy Shaya<sup>4</sup>, Monique Topp<sup>4</sup>, Tianyao Lu<sup>3</sup>, Robert Tobler<sup>3</sup>, Adam L. Valkovic<sup>3</sup>, Zachery Moore<sup>3,5</sup>, Oluwaseun E. Fatunla<sup>3,5</sup>, Jurgen Kriel<sup>3,5</sup>, Joel J. D. Moffet<sup>3,5</sup>, Heidi McAlpine<sup>1,6</sup>, Marius Rosiersup<sup>6</sup>, Hefei Guan<sup>6</sup>, James Dimou<sup>1,2</sup>, Verena Schadewaldt<sup>1,2</sup>, Samuel Roberts-Thomson<sup>7</sup>, David McArdle<sup>8,9</sup>, Elaine Lui<sup>8,9</sup>, Moritz Voelker-Albert<sup>10</sup>, Simone di Sanzo<sup>10</sup>, Brunda Nijagal<sup>11</sup>, Vinod K. Narayana<sup>10</sup>, Camilla B. Mitchell<sup>12,13</sup>, Joseph Vissers<sup>12,13</sup>, Lucy Palmer<sup>6</sup>, Sean Grimmond<sup>13</sup>, Mark A. Rosenthal<sup>4</sup>, Sarah A. Best<sup>3,5*</sup>, Saskia Freytag<sup>3,5*</sup>, James R. Whittle<sup>3,4,5,#,*</sup>.  

Affiliations 
1. Department of Neurosurgery, Royal Melbourne Hospital, Parkville, Australia, 3050
2. Department of Surgery, University of Melbourne, Parkville, Australia, 3052
3. Personalised Oncology Division, WEHI, Parkville, Australia, 3052
4. Department of Medical Oncology, Peter MacCallum Cancer Centre, Melbourne, Australia, 3000
5. Department of Medical Biology, University of Melbourne, Parkville, Australia, 3052
6. The Florey Institute of Neuroscience and Mental Health, The Florey Department of Neuroscience and Mental Health, University of Melbourne, Parkville, Australia, 3052
7. Department of Anatomical Pathology, Royal Melbourne Hospital, Parkville, Australia, 3050
8. Department of Medical Imaging, Royal Melbourne Hospital, Parkville, Australia, 3050
9. Department of Radiology, University of Melbourne, Parkville, Australia, 3052
10. MoleQlar Analytics GmbH, Munich, Germany, 81671
11. Metabolomics Australia, Bio21 Molecular Science and Biotechnology Institute, The University of Melbourne, Melbourne, Australia, 3010
12. Department Clinical Pathology, MDHS, University of Melbourne, Melbourne, Australia, 3010
13. Collaborative Centre for Genomic Cancer Medicine, MDHS, University of Melbourne, Melbourne, Australia, 3010  

*    These authors jointly supervised this work  

(# Co-corresponding authors) 
(* Shared senior authorship)  

Correspondence 
 
Dr Jim Whittle 
jim.whittle@petermac.org 
 
Prof. Kate Drummond 
kate.drummond@mh.org.au 

All original code pertaining to analysis for the study is organised in the following multi-omics directories:
### - Metabolomics: Bulk metabolomics
### - Spatial_Metabolomics: Spatial metabolomics
### - Spatial_Transcriptomics: 10X Xenium spatial transcriptomics
### - snRNA-seq: 10X Chromium single nuclei RNA sequencing (5' single-end)
### - Methylation_Array: Infinium MethylationEPIC v2.0 BeadChip array
### - Epigenetic: Histone Modifications LC-MS
### - Immunofluorescence

Data Availability:
Study‐level clinical and spatial data from this study will be made available upon reasonable request to the lead authors from a qualified medical or scientific professional for the specific purpose laid out in that request and may include deidentified individual participant data (requests evaluated within 4 weeks). The data for this request will be available after a data access agreement has been signed. Further translational data is available from dbGAP under accession code phs003976 after provision of IRB ethics approval and the snRNA-seq data for the independent LGG cohort is available at GEO under accession code GSE292732. Data for the matched spatial transcriptomics and metabolomics data from sequential sections of the same LGG tumor is available at BioImage Archive via DOI: 10.6019/S-BIAD1426.  

Publicly available data from Spitzer et al. 2024 is available at GEO under the accession number GSE26099 under the terms and conditions specified by the data provider. Single-cell sequencing reads were aligned and quantified using the GRCh38 human reference genome (2020-A), a pre-built reference package provided by 10x Genomics (GENCODEv32/Ensembl98). The reference package is available at: https://www.10xgenomics.com/support/software/cell-ranger/latest/release-notes/cr-reference-release-notes#2020-a. 

WGS reads were aligned using the hg38 reference available at DRAGEN Illumina: https://sapac.support.illumina.com/downloads/dragen-reference-genomes-hg38.html. 

Any additional information required to reanalyze the data reported in the study is available from the lead contact upon request. 
