---
title: "Load & Filter Methylation (EPIC v2) Array Data"
output: html_document
---

The following code relates to the loading and filtering of all participant raw methylation array data (idat files)

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(readxl)
library(minfi)
library(limma)
library(IlluminaHumanMethylationEPICanno.ilm10b4.hg19)
library(IlluminaHumanMethylationEPICmanifest)
library(RColorBrewer)
library(missMethyl)
library(Gviz)
library(stringr)
library(ggplot2)
library(parallel)
```

## Load Annotation Data 

```{r}
# EPIC annotation
epic <- getAnnotation(IlluminaHumanMethylationEPICanno.ilm10b4.hg19)
head(epic)
```

## Load methylation array data

```{r}
# directory of all idat files
dataDirectory <- "/data_raw/MTH/"

# read in sample sheet
targets <- read.metharray.sheet(dataDirectory, pattern="SampleSheet.csv")
targets$Sample_Type <- rep(c("Pre-treatment", "Post-treatment"), times = length(targets$Sample) / 2)
targets$Patient <- sub("([^_]+).*", "\\1", targets$Sample)
targets

# read in raw data red/green IDAT files
rgSet <- read.metharray.exp(targets=targets)
sampleNames(rgSet) <- targets$Sample
rgSet@annotation <- c(array='IlluminaHumanMethylationEPIC', annotation='ilm10b4.hg19')
rgSet
```

## Quality Control

```{r}
detP <- detectionP(rgSet)
head(detP)

pal <- Polychrome::palette36.colors(4)
par(mfrow=c(1,2))
barplot(colMeans(detP), col=pal[factor(targets$Sample)], las=2, 
        cex.names=0.6, ylab="Mean detection p-values")
abline(h=0.05,col="red")

#qcReport(rgSet, sampNames=targets$Sample, 
#                  pdf="figures/qcReport.pdf")
```

## Filter Samples

```{r}
keep <- colMeans(detP) < 0.05
rgSet <- rgSet[,keep]
rgSet

targets <- targets[keep,]
targets[,1:5]

detP <- detP[,keep]
dim(detP)
```
No samples filtered.

## Filter Outlier Peaks

```{r}
mSetRaw <- preprocessRaw(rgSet)
densityPlot(getBeta(mSetRaw), main="Raw", legend=FALSE)

find_peaks <- function(x) {
  d <- density(x[!is.na(x)])
  d$x[c(F, diff(diff(d$y)>=0)<0)]
}

all_peaks <- apply(getBeta(mSetRaw), 2,function(x) 
  find_peaks(x))
table(lengths(all_peaks)>3)
```
No outlier peaks detected.

## Normalisation

```{r}
mSetSq <- preprocessQuantile(rgSet) 

layout(matrix(c(1,1,3,2,2,3), 2, 3, byrow = TRUE), 
       widths = c(1, 1, 0.3), heights = c(1, 1))
densityPlot(rgSet, sampGroups=targets$Sample, main="Raw", legend=FALSE, pal = pal)
densityPlot(getBeta(mSetSq), sampGroups=targets$Sample, main="Normalized", legend=FALSE, pal = pal)
par(mar = c(0, 0, 0, 0))  
plot.new() 
legend("center", legend = levels(factor(targets$Sample)), text.col = pal, bty = "n")
```

## MDS Plot

```{r}
par(mfrow=c(1,2))
plotMDS(getM(mSetSq), top=1000, gene.selection="common", 
        col=pal[factor(targets$Sample)], cex=0.9, pch = 19)
legend("top", legend=levels(factor(targets$Sample)), text.col=pal,
       bg="white", cex=0.4)

plotMDS(getM(mSetSq), top=1000, gene.selection="common",  
        col=pal[factor(targets$Sample_Type)], cex=0.9, pch = 19)
legend("top", legend=levels(factor(targets$Sample_Type)), text.col=pal,
       bg="white", cex=0.4)

# higher dimensions
par(mfrow=c(1,3))
plotMDS(getM(mSetSq), top=1000, gene.selection="common", 
        col=pal[factor(targets$Sample)], dim=c(1,3), pch = 19)
legend("top", legend=levels(factor(targets$Sample)), text.col=pal, 
       cex=0.7, bg="white")

plotMDS(getM(mSetSq), top=1000, gene.selection="common", 
        col=pal[factor(targets$Sample)], dim=c(2,3), pch = 19)

plotMDS(getM(mSetSq), top=1000, gene.selection="common", 
        col=pal[factor(targets$Sample)], dim=c(3,4), pch = 19)
```

## Filter Failed/Poor Quality Probes

```{r}
# Ensure probes are in the same order
detP <- detP[match(featureNames(mSetSq),rownames(detP)),] 

keep <- rowSums(detP < 0.01) == ncol(mSetSq) 
table(keep)

mSetSqFlt <- mSetSq[keep,]
mSetSqFlt
```

## Filter CGs and Sex Chromosomes

```{r}
keep <- !(featureNames(mSetSqFlt) %in% epic$Name[epic$chr %in% 
                                                        c("chrX","chrY")])
table(keep)

mSetSqFlt <- mSetSqFlt[keep,]
```

## Filter CGs with SNPs and potential cross-reactivity

```{r}
mSetSqFlt <- mapToGenome(mSetSqFlt)
mSetSqFlt <- dropLociWithSnps(mSetSqFlt)
mSetSqFlt
```

```{r}
xReactiveProbes <- read_xlsx("data/download/48639-non-specific-probes-Illumina450k.xlsx")
keep <- !(featureNames(mSetSqFlt) %in% xReactiveProbes$TargetID)
table(keep)

mSetSqFlt <- mSetSqFlt[keep,] 
mSetSqFlt
```

```{r}
# re-examine MDS
par(mfrow=c(1,2))
plotMDS(getM(mSetSqFlt), top=1000, gene.selection="common", 
        col=pal[factor(targets$Sample)], cex=0.8, pch = 19)
legend("top", legend=levels(factor(targets$Sample)), text.col=pal,
       cex=0.3, bg="white")

plotMDS(getM(mSetSqFlt), top=1000, gene.selection="common", 
        col=pal[factor(targets$Sample_Type)], cex=0.8, pch = 19)
legend("top", legend=levels(factor(targets$Sample_Type)), text.col=pal,
       cex=0.7, bg="white")
```

## Save filtered and processed files

```{r}
#saveRDS(mSetSqFlt, file= "data/process/filtered_data.rds")
#saveRDS(targets, file = "data/process/targets.rds")
```
